compiler/src/iree/compiler/Codegen/Common/OptimizeTensorInsertExtractSlices.cpp

```diff
@@ -36,6 +37,169 @@ class OptimizeTensorInsertExtractSlicesPass
   void runOnOperation() override;
 };
 
+/// Checks whether the given op can be hoisted by checking that
+/// - the op and none of its contained operations depend on values inside of the
+///   loop (by means of calling definedOutside).
+/// - the op has no side-effects.
+static bool canBeHoisted(Operation *op,
+                         function_ref<bool(OpOperand &)> condition) {
+  // Do not move terminators.
+  if (op->hasTrait<OpTrait::IsTerminator>())
+    return false;
+
+  // Walk the nested operations and check that all used values are either
+  // defined outside of the loop or in a nested region, but not at the level of
+  // the loop body.
+  auto walkFn = [&](Operation *child) {
+    for (OpOperand &operand : child->getOpOperands()) {
+      // Ignore values defined in a nested region.
+      if (op->isAncestor(operand.get().getParentRegion()->getParentOp()))
+        continue;
+      if (!condition(operand))
+        return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  };
+  return !op->walk(walkFn).wasInterrupted();
+}
+
+/// Return the newly created loop op (that has extra iter_args) or the original
+/// loop op if nothing was hoisted.
+static LoopLikeOpInterface
+hoistLoopInvariantSubsetAtIterArg(RewriterBase &rewriter,
+                                  LoopLikeOpInterface loopLike, int64_t idx) {
+  // Get subset insertion of this yielded arg.
+  auto insertion = loopLike.getYieldedValues()[idx]
+                       .getDefiningOp<SubsetInsertionOpInterface>();
+  if (!insertion) {
+    return loopLike;
+  }
+  // Check if this insertion is loop invariant except it's source.
+  // We would also be okay as long as the destination is loop invariant,
+  // but we would have to do some cloning, so we don't do it here.
+  if (!canBeHoisted(insertion, [&](OpOperand &operand) {
+        return loopLike.isDefinedOutsideOfLoop(operand.get()) ||
+               &operand == &insertion.getSourceOperand();
+      })) {
+    return loopLike;
+  }
+
+  // Get all subset extraction uses of this iter_arg and try to hoist them
+  // out of the loop.
+  for (Operation *op : loopLike.getRegionIterArgs()[idx].getUsers()) {
+    auto extraction = dyn_cast<SubsetExtractionOpInterface>(op);
+    if (!extraction) {
+      continue;
+    }
+
+    // Check if this extraction is operating on the same subset as the
+    // insertion.
+    bool equivalent = extraction.operatesOnEquivalentSubset(
```

Comment:
Tried to look at the documentation of this method. I have no idea what the call back does.

Reply:
I couldnt quite follow this method, cause without making sure you are extracting/inserting from/to the same tensor source/dest, what is the point of comparing the insert and extract slices.... This seems like an artificial coupling.

Reply:
The coupling is happening through iter args. So just checking the slices extracted and inserted is enough. They might actually be to different tensors

Reply:
We discussed this offline, but for future documentation, this get's implicitly checked. The result of insert_slice is the yielded value and the user of the corresponding iter_arg is the extract_slice.